ank-core/src/main/kotlin/io/kategory/ank/interpreter.kt

```diff
@@ -118,37 +126,58 @@ fun extractCodeImpl(source: String, tree: ASTNode): ListKW<Snippet> {
     return sb.k()
 }
 
-fun compileCodeImpl(origin: File, snippets: ListKW<Snippet>, classpath: ListKW<String>): CompiledMarkdown {
-    println(":runAnk -> Compiling: $origin")
-    val classLoader = URLClassLoader(classpath.map { URL(it) }.ev().list.toTypedArray())
-    val seManager = ScriptEngineManager(classLoader)
-    val cachedEngines: ConcurrentHashMap<String, ScriptEngine> = ConcurrentHashMap()
-    val evaledSnippets = snippets.list.map { snippet ->
-        val result = Try {
-            val engine = cachedEngines.get(snippet.lang)
-            val resolvedEngine = if (engine != null) engine else {
-                seManager.getEngineByExtension(extensionMappings.getOrDefault(snippet.lang, "kts"))!!
+
+fun compileCodeImpl(snippets: Map<File, ListKW<Snippet>>, classpath: ListKW<String>): ListKW<CompiledMarkdown> {
+    println(":runAnk -> started compilation")
+    val IOPool = object : CoroutineDispatcher() {
+        private val ioBoundPool = Executors.newFixedThreadPool(snippets.size)
+        override fun dispatch(context: CoroutineContext, block: Runnable) {
+            ioBoundPool.submit(block)
+        }
+    }
+    return runBlocking {
+        snippets.map { (file, codeBlocks) ->
+            async(IOPool) {
+                val start = System.currentTimeMillis()
+                val classLoader = URLClassLoader(classpath.map { URL(it) }.ev().list.toTypedArray())
+                val seManager = ScriptEngineManager(classLoader)
+                println(":runAnk -> START eval: $file on ${Thread.currentThread().name}")
+                val engineCache: Map<String, ScriptEngine> =
```

Comment:
Should the engine cache be local per script, or shared among all of them?

Reply:
I was told they were thread safe but those rx issues I mentioned to you in private show they are not so there needs to be one instance per file basically. Those engines are not thread safe.